src/NLog/Config/ServiceRepository.cs

```diff
@@ -95,30 +97,70 @@ private object DefaultResolveInstance(Type itemType, HashSet<Type> seenTypes)
                 return objectResolver(itemType);
             }
 
-            try
+            if (_lateBoundMap.TryGetValue(itemType, out var compiledConstructor))
             {
-                //todo DI cache/find upfront
-                var defaultConstructor = itemType.GetConstructor(Type.EmptyTypes);
-                if (defaultConstructor != null)
-                {
-                    InternalLogger.Trace("Found public default ctor");
-                    return defaultConstructor.Invoke(ArrayHelper.Empty<object>());
-                }
-
-                seenTypes = seenTypes ?? new HashSet<Type>();
-                return CreateFromParameterizedConstructor(itemType, seenTypes);
+                return CreateNewInstance(compiledConstructor, seenTypes);
             }
-            catch (MissingMethodException exception)
+
+            //todo? lock (_lateBoundMapLock)
```

Comment:
Uncomment the code and extract the locked-section into separate method. When having taken the lock, then re-check _lateBoundMap if Type has been inserted (TryGetValue-again)

Reply:
done :)

please double check, thanks!